File: test/fixtures/golden/multi-target/formula.rb
class ReleaseExampleMultiTarget < Formula
  desc "release-example-multi-target 0.1.0 release artifact"
  homepage "https://github.com/owner/release-example-multi-target"
  url "https://github.com/owner/release-example-multi-target/releases/download/v0.1.0/release-example-multi-target-0.1.0.tgz"
  sha256 "60bf2ce69a7cd8d437fa36701f8913bfe92a9f37aedea8e191f34fc6d11f9c4b"
  version "0.1.0"

  def install
    bin.install "bin/release-example-multi-target" => "release-example-multi-target"
    chmod 0755, bin/"release-example-multi-target"
  end

  test do
    assert File.exist?(bin/"release-example-multi-target")
    assert File.executable?(bin/"release-example-multi-target")
  end
end

File: packages/catalog/src/homebrew/Formula.ts
import { Effect, Schema } from "effect"
import { File, type Bundle } from "@mannyc1/ts-release/bundle"
import { Text, Url, Executable, decode, downloads, renderBytes } from "../Shared.js"

export class Download extends Schema.Class<Download>("Homebrew.Download")({
  url: Url,
  file: File,
}) {}
export class Formula extends Schema.Class<Formula>("Homebrew.Formula")({
  className: Text.check(
    Schema.isPattern(/^[A-Z][A-Za-z0-9]*$/u),
    Schema.makeFilter((value) => value !== "BEGIN" && value !== "END"),
  ),
  description: Text,
  homepage: Url,
  license: Text,
  version: Text,
  executable: Executable,
  archives: Schema.Struct({
    "darwin-x64": Download,
    "darwin-arm64": Download,
    "linux-x64": Download,
    "linux-arm64": Download,
  }),
}) {}
// Ruby double-quoted strings interpolate #{}, #@ and #$; JSON quoting alone is unsafe.
const ruby = (value: string) => JSON.stringify(value).replace(/#(?=[{@$])/gu, "\\#")
export const render = Effect.fn("Homebrew.render")((input: Formula, bundle: Bundle) =>
  renderBytes(() => {
    const value = decode(Formula, input)
    downloads(bundle, value.archives)
    const lines = [
      `class ${value.className} < Formula`,
      `  desc ${ruby(value.description)}`,
      `  homepage ${ruby(value.homepage)}`,
      `  license ${ruby(value.license)}`,
    ]
    for (const os of ["darwin", "linux"] as const) {
      lines.push("", `  on_${os === "darwin" ? "macos" : "linux"} do`)
      for (const arch of ["arm64", "x64"] as const) {
        const download = value.archives[`${os}-${arch}`]
        lines.push(
          `    on_${arch === "arm64" ? "arm" : "intel"} do`,
          `      url ${ruby(download.url)}`,
          `      sha256 ${ruby(download.file.content.sha256)}`,
          "    end",
        )
      }
      lines.push("  end")
    }
    lines.push(
      "",
      `  version ${ruby(value.version)} if version.to_s != ${ruby(value.version)}`,
      "",
      "  def install",
      `    bin.install ${ruby(value.executable)}`,
      "  end",
      "",
      "  test do",
      `    assert_predicate bin/${ruby(value.executable.split("/").at(-1)!)}, :executable?`,
      "  end",
      "end",
      "",
    )
    return lines.join("\n")
  }),
)
